In [17]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

Now it's your turn. Using the space below, experiment with different hidden layer structures. You can try this on a subset of the data to improve runtime. See how things vary. See what seems to matter the most. Feel free to manipulate other parameters as well. It may also be beneficial to do some real feature selection work...

In [18]:
artworks = pd.read_csv('https://media.githubusercontent.com/media/MuseumofModernArt/collection/master/Artworks.csv')

In [19]:
artworks.columns

Index(['Title', 'Artist', 'ConstituentID', 'ArtistBio', 'Nationality',
       'BeginDate', 'EndDate', 'Gender', 'Date', 'Medium', 'Dimensions',
       'CreditLine', 'AccessionNumber', 'Classification', 'Department',
       'DateAcquired', 'Cataloged', 'ObjectID', 'URL', 'ThumbnailURL',
       'Circumference (cm)', 'Depth (cm)', 'Diameter (cm)', 'Height (cm)',
       'Length (cm)', 'Weight (kg)', 'Width (cm)', 'Seat Height (cm)',
       'Duration (sec.)'],
      dtype='object')

We'll also do a bit of data processing and cleaning, selecting columns of interest and converting URL's to booleans indicating whether they are present.

In [20]:
# Select Columns.
artworks = artworks[['Artist', 'Nationality', 'Gender', 'Date', 'Department',
                    'DateAcquired', 'URL', 'ThumbnailURL', 'Height (cm)', 'Width (cm)']]

# Convert URL's to booleans.
artworks['URL'] = artworks['URL'].notnull()
artworks['ThumbnailURL'] = artworks['ThumbnailURL'].notnull()

# Drop films and some other tricky rows.
artworks = artworks[artworks['Department']!='Film']
artworks = artworks[artworks['Department']!='Media and Performance Art']
artworks = artworks[artworks['Department']!='Fluxus Collection']

# Drop missing data.
artworks = artworks.dropna()

In [21]:
artworks.head()

Artist Nationality  Gender  Date             Department  \
0               Otto Wagner  (Austrian)  (Male)  1896  Architecture & Design   
1  Christian de Portzamparc    (French)  (Male)  1987  Architecture & Design   
2                Emil Hoppe  (Austrian)  (Male)  1903  Architecture & Design   
3           Bernard Tschumi          ()  (Male)  1980  Architecture & Design   
4                Emil Hoppe  (Austrian)  (Male)  1903  Architecture & Design   

  DateAcquired   URL  ThumbnailURL  Height (cm)  Width (cm)  
0   1996-04-09  True          True      48.6000    168.9000  
1   1995-01-17  True          True      40.6401     29.8451  
2   1997-01-15  True          True      34.3000     31.8000  
3   1995-01-17  True          True      50.8000     50.8000  
4   1997-01-15  True          True      38.4000     19.1000

Building a Model
Now, let's see if we can use multi-layer perceptron modeling (or "MLP") to see if we can classify the department a piece should go into using everything but the department name.

Before we import MLP from SKLearn and establish the model we first have to ensure correct typing for our data and do some other cleaning

In [22]:
# Get data types.
artworks.dtypes

Artist           object
Nationality      object
Gender           object
Date             object
Department       object
DateAcquired     object
URL                bool
ThumbnailURL       bool
Height (cm)     float64
Width (cm)      float64
dtype: object

The DateAcquired column is an object. Let's transform that to a datetime object and add a feature for just the year the artwork was acquired.

In [23]:
artworks['DateAcquired'] = pd.to_datetime(artworks.DateAcquired)
artworks['YearAcquired'] = artworks.DateAcquired.dt.year
artworks['YearAcquired'].dtype

dtype('int64')

Great. Let's do some more miscellaneous cleaning.

In [24]:
# Remove multiple nationalities, genders, and artists.
artworks.loc[artworks['Gender'].str.contains('\) \('), 'Gender'] = '\(multiple_persons\)'
artworks.loc[artworks['Nationality'].str.contains('\) \('), 'Nationality'] = '\(multiple_nationalities\)'
artworks.loc[artworks['Artist'].str.contains(','), 'Artist'] = 'Multiple_Artists'

# Convert dates to start date, cutting down number of distinct examples.
artworks['Date'] = pd.Series(artworks.Date.str.extract(
    '([0-9]{4})', expand=False))[:-1]

# Final column drops and NA drop.
X = artworks.drop(['Department', 'DateAcquired', 'Artist', 'Nationality', 'Date', 'URL', 'ThumbnailURL'], 1)

# Create dummies separately.
artists = pd.get_dummies(artworks.Artist)
nationalities = pd.get_dummies(artworks.Nationality)
dates = pd.get_dummies(artworks.Date)

# Concat with other variables, but artists slows this wayyyyy down so we'll keep it out for now
X = pd.get_dummies(X, sparse=True)
X = pd.concat([X, nationalities, dates], axis=1)

Y = artworks.Department

In [25]:
artworks.head()

Artist Nationality  Gender  Date             Department  \
0               Otto Wagner  (Austrian)  (Male)  1896  Architecture & Design   
1  Christian de Portzamparc    (French)  (Male)  1987  Architecture & Design   
2                Emil Hoppe  (Austrian)  (Male)  1903  Architecture & Design   
3           Bernard Tschumi          ()  (Male)  1980  Architecture & Design   
4                Emil Hoppe  (Austrian)  (Male)  1903  Architecture & Design   

  DateAcquired   URL  ThumbnailURL  Height (cm)  Width (cm)  YearAcquired  
0   1996-04-09  True          True      48.6000    168.9000          1996  
1   1995-01-17  True          True      40.6401     29.8451          1995  
2   1997-01-15  True          True      34.3000     31.8000          1997  
3   1995-01-17  True          True      50.8000     50.8000          1995  
4   1997-01-15  True          True      38.4000     19.1000          1997

In [26]:
# Import the model.
from sklearn.neural_network import MLPClassifier

# Establish and fit the model, with a single, 100 perceptron layer.
mlp = MLPClassifier(hidden_layer_sizes=(100,))
mlp.fit(X, Y)

C:\Users\18644\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [27]:
mlp.score(X, Y)

0.7779704157290936

In [28]:
Y.value_counts()/len(Y)

Drawings & Prints        0.622496
Photography              0.226428
Architecture & Design    0.113016
Painting & Sculpture     0.033688
Media and Performance    0.004371
Name: Department, dtype: float64

In [31]:
from sklearn.model_selection import cross_val_score
cross_val = cross_val_score(mlp, X, Y, cv=5)
print('Cross Validation Score: {:.3f}(+/-{:.2f})'.format(cross_val.mean(), cross_val.std()*2))

Cross Validation Score: 0.657(+/-0.11)


The MLP score is higher(78% approx) as it's based on the entire dataset whereas cross val has 5 scores and on average it's lower (66% approx)as cross validation would split data into 5 folds and use one of them each time to val and the rest for fitting the model.
Hence the cross validation scores are the accuracy based on unknown dataset model hasn't seen before.
The average score is 0.65 (66% approx) with some variance. It suggests there might be a bit overfit and we can reduce the hidden layer and see if the model generalize better. 
